File: bill_buddy_backend/src/model/payment.js
const mongoose = require('mongoose');

const paymentSchema = new mongoose.Schema(
  {
    userId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true,
    },

    billId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Bill',
      required: true,
    },

    amount: {
      type: Number,
      required: true,
      min: 0,
    },

    status: {
      type: String,
      enum: ['success', 'pending', 'failed'],
      default: 'success',
    },

    transactionId: {
      type: String,
      required: true,
      unique: true,
    },

    idempotencyKey: {
      type: String,
      required: true,
      unique: true,
    },

    paidAt: {
      type: Date,
      default: Date.now,
    },
  },
  {
    timestamps: true,
  },
);

module.exports = mongoose.model(
  'Payment',
  paymentSchema,
  'payments',
);

File: bill_buddy/lib/models/payment.dart
class Payment {
  final String id;
  final String billId;
  final double amount;
  final String status;
  final String transactionId;
  final DateTime paidAt;

  const Payment({
    required this.id,
    required this.billId,
    required this.amount,
    required this.status,
    required this.transactionId,
    required this.paidAt,
  });

  factory Payment.fromJson(Map<String, dynamic> json) {
    return Payment(
      id: json['_id'] as String,
      billId: json['billId'] as String,
      amount: (json['amount'] as num).toDouble(),
      status: json['status'] as String,
      transactionId: json['transactionId'] as String,
      paidAt: DateTime.parse(json['paidAt'] as String),
    );
  }
}